gsy_framework/sim_results/electric_blue/aggregate_results.py

```diff
@@ -0,0 +1,145 @@
+        assert self.device_uuid == other.device_uuid
+
+        if self.current_timeslot >= other.current_timeslot:
+            current_timeslot = self.current_timeslot
+            open_bids = self.open_bids
+            open_offers = self.open_offers
+        else:
+            current_timeslot = other.current_timeslot
+            open_bids = other.open_bids
+            open_offers = other.open_offers
+
+        return ForwardDeviceStats(
+            timeslot=self.timeslot,
+            device_uuid=self.device_uuid,
+            current_timeslot=current_timeslot,
+            total_energy_produced=self.total_energy_produced + other.total_energy_produced,
+            total_sell_trade_count=self.total_sell_trade_count + other.total_sell_trade_count,
+            total_energy_sold=self.total_energy_sold + other.total_energy_sold,
+            total_earned_eur=self.total_earned_eur + other.total_earned_eur,
+            total_energy_consumed=self.total_energy_consumed + other.total_energy_consumed,
+            total_buy_trade_count=self.total_buy_trade_count + other.total_buy_trade_count,
+            total_energy_bought=self.total_energy_bought + other.total_energy_bought,
+            total_spent_eur=self.total_spent_eur + other.total_spent_eur,
+            open_bids=open_bids,
+            open_offers=open_offers,
+            trades=self.trades + other.trades
+        )
+
+    def to_dict(self) -> Dict:
+        """Generate a dictionary for saving the data into DB."""
+        return {f.name: getattr(self, f.name) for f in fields(self)}
+
+    def add_trade(self, trade: Dict) -> None:
+        """Add trade information to device stats."""
+        assert trade["timeslot"] == self.timeslot
+
+        if trade["seller_id"] == self.device_uuid:
+            self.trades.append(trade)
+            self.total_sell_trade_count += 1
+            self.total_energy_sold += trade["energy"]
+            self.total_earned_eur += trade["price"]
+        elif trade["buyer_id"] == self.device_uuid:
+            self.trades.append(trade)
+            self.total_buy_trade_count += 1
+            self.total_energy_bought += trade["energy"]
+            self.total_spent_eur += trade["price"]
+        else:
+            raise AssertionError("Device is not seller/buyer of the trade.")
+
+    def add_bid(self, bid: Dict) -> None:
+        """Add bid information to device stats."""
+        assert bid["timeslot"] == self.timeslot
+        assert bid["buyer_id"] == self.device_uuid, "Device is not buyer of the bid."
+        self.open_bids.append(bid)
+
+    def add_offer(self, offer: Dict) -> None:
+        """Add offer information to device stats."""
+        assert offer["timeslot"] == self.timeslot
+        assert offer["seller_id"] == self.device_uuid, "Device is not seller of the offer."
+        self.open_offers.append(offer)
+
+
+def handle_forward_results(
+        current_timeslot: DateTime, market_stats: Dict[str, Dict[str, List]]) -> Dict:
```

Comment:
I understand that `market_stats` are coming from the gsy-e and contain the latest orders, meaning the orders that are performed in the forward markets, right?
Can you please sketch the content of the `market_stats` dict here? Is it the following?
```
market_stats = {
    "20220101T00:00": {"offers": [], "bids": [], "trades": []},
    ...
}
```
?
What about the trades that happened in the same market slots but are not part of the `market_slots` any more. 
Are their buffered somewhere? Or is there no need?
I am asking because you are performing `self.total_energy_sold += trade["energy"]` in `ForwardDeviceStats.add_trade()`. So the result of this function should also be input in the next market slot, right?
Or am I missing something here?

Reply:
The market_stats is exactly like what you mentioned.
Also, trades of the same forward slot that have happened in the previous slot are buffered in `ForwardDeviceStats.trades`. That's exactly why we need to sum the previous ForwardDeviceStats with the current one. In other words, the result of the last slot should be feeded into the next one.

Reply:
But I do not see, where the `why we need to sum the previous ForwardDeviceStats with the current one` is done. In `handle_forward_results` the `results` is an empty dict at the beginning of the method. Can you please help me understand where the previous results are buffered? Thanks!

Reply:
Adding the previous results to the current results is not done here. It will be done by @cruky. In order to do that:
1- Fetch the last results from DB
2- Create the ForwardDeviceStats object from last results
3- Using the __add__ magic method, he will be able to do: last_results + current_results
4- Save the updated results into the DB to be used for next market slot.

I only provide the __add__ method so that Kamil can easily accumulate the results.

Reply:
Thanks a lot for explaining!

Reply:
The `handle_forward_results` only creates the current results. Handling the last results would be another ticket in the future.